src/JWTVerifier.php

```diff
@@ -177,53 +178,35 @@ public function verifyAndDecode($jwt)
         }
 
         try {
-            $head_decoded = $this->decodeTokenSegment($tks[0]);
             $body_decoded = $this->decodeTokenSegment($tks[1]);
         } catch (\DomainException $e) {
             throw new InvalidTokenException('Malformed token.');
         }
 
-        if (! is_object($head_decoded) || ! is_object($body_decoded)) {
-            throw new InvalidTokenException('Malformed token.');
-        }
-
-        if (empty($head_decoded->alg)) {
-            throw new InvalidTokenException('Token algorithm not found');
-        }
-
-        if (! $this->supportsAlg($head_decoded->alg)) {
-            throw new InvalidTokenException('Token algorithm not supported');
-        }
-
         // Validate the token audience, if present.
         if (! empty($body_decoded->aud)) {
             $audience = is_array($body_decoded->aud) ? $body_decoded->aud : [$body_decoded->aud];
             if (! count(array_intersect($audience, $this->valid_audiences))) {
-                $message  = 'Invalid token audience '.implode( ', ', $audience );
-                $message .= '; expected '.implode( ', ', $this->valid_audiences );
-                throw new InvalidTokenException($message);
+                throw new InvalidTokenException('Invalid token audience');
             }
         }
 
-        if ('HS256' === $head_decoded->alg) {
-            $secret = $this->client_secret;
+        // At this point, the constructor has set the supported algs to ensure only one.
+        if ($this->supportsAlg( 'HS256' )) {
+            $signature_key = $this->client_secret;
         } else {
-            if (empty($head_decoded->kid)) {
-                throw new CoreException('Token key ID is missing for RS256 token');
-            }
-
             if (empty($body_decoded->iss) || ! in_array($body_decoded->iss, $this->authorized_iss)) {
```

Comment:
issuer check should be done for HS256 case as well. And AFTER verifying the signature. (just like I asked for the audience)

Reply:
Requiring an issuer for HS256 would be another breaking change in this class. Should I proceed?

Reply:
I moved the check to after the signature validation and this will pick up checks if valid issuers have been added to the config. I can't force an issuer being added for HS256 tokens to this class without another breaking change (though the helper function that calls this always adds one). Should I add this additional break in this PR or leave it as-is for now?